---
title: "HMMRATAC Guide"
author: "Daniel Van Twisk, Evan Tarbell"
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('HMMRATAC')`"
bibliography: bibliography.bib
csl: nature.csl
vignette: >
  %\VignetteIndexEntry{HMMRATAC Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document
---

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(ATACseqQC)
  library(ChIPpeakAnno)
  library(BSgenome.Hsapiens.UCSC.hg19)
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  library(phastCons100way.UCSC.hg19)
  library(MotifDb)
  library(GenomicAlignments)
})
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

# Introduction
Assay for Transposase-Accessible Chromatin using sequencing (ATAC-seq) is a
method to identify open chromatin regions in a genome-wide manner. This method
involves using a transposase to cut sections of accessable DNA selectively [@buenostro2013transposition].
ATAC-seq identifies regions highly corrrelated with similar methods used to
identify these regions, DNase-seq [@boyle2008high] and FAIRE-seq [@giresi2007faire]. ATAC-seq has become a popular
method for this task because it is faster and easier to perform, has a higher
signal to noise ration, and does not require as large of an amount of biological
samples when comapred with other methods [@buenostro2013transposition].

Hidden Markov ModeleR for ATAC-seq (HMMRATAC) is a semi-supervised machine
learning appraoch for identifying open chromatin regions from ATAC-seq data.
The idea behind HMMRATAC is conceptualized upon the idea of "decomposition and
integration. The ATAC-seq data is decomposed into different layers of coverage
signals corresponding to the sequnced DNA fragments originated from nucleosomal
regions. The relationship between the layers of signals at open chromatin
regions is then learned using a Hidden Markov Model and utilized for predicting
open chromatin. HMMRATAC is capable of outperforming similar methods in
identifying chromatin structure and transcription factor binding sites. 
[@hmmratac2019tarbell] 

A typical analysis pipline begins with aligning sequencing reads to a reference
genome. 

This Bioconductor package has been made to utilize the original Java program for
HMMRATAC and make it availiable for use within Bioconductor.

# Installation
```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install('HMMRATAC')
```

# Example Pipeline

We start by first obtaining a list of 

## HMMRATAC

First off, HMMRATAC requires a large allocation of memory to run that rJava
is not able to initially allocate to it. The amount of memory needed varies by
the size of the data being modeled. For this example, we allocate 8GB of memory
before loading the HMMRATAC package:
```{r loadHMMRATAC}
options(java.parameters = "-Xmx8000m")
library(HMMRATAC)
```

### HMMR output files

The following files are generated in an output folder from a run of HMMRATAC:


# Session Info
```{r sessionInfo}
sessionInfo()
```
