---
title: "HMMRATAC Guide"
author: "Daniel Van Twisk, Evan Tarbell"
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('HMMRATAC')`"
bibliography: bibliography.bib
csl: nature.csl
vignette: >
  %\VignetteIndexEntry{HMMRATAC Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document
---

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(ATACseqQC)
  library(ChIPpeakAnno)
  library(BSgenome.Hsapiens.UCSC.hg19)
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  library(phastCons100way.UCSC.hg19)
  library(MotifDb)
  library(GenomicAlignments)
})
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

# Introduction
Assay for Transposase-Accessible Chromatin using sequencing (ATAC-seq) is a
method to identify open chromatin regions in a genome-wide manner. This method
involves using a transposase to cut sections of accessable DNA selectively [@buenostro2013transposition].
ATAC-seq identifies regions highly corrrelated with similar methods used to
identify these regions, DNase-seq [@boyle2008high] and FAIRE-seq [@giresi2007faire]. ATAC-seq has become a popular
method for this task because it is faster and easier to perform, has a higher
signal to noise ration, and does not require as large of an amount of biological
samples when comapred with other methods [@buenostro2013transposition].

Hidden Markov ModeleR for ATAC-seq (HMMRATAC) is a semi-supervised machine
learning appraoch for identifying open chromatin regions from ATAC-seq data.
The idea behind HMMRATAC is conceptualized upon the idea of "decomposition and
integration. The ATAC-seq data is decomposed into different layers of coverage
signals corresponding to the sequnced DNA fragments originated from nucleosomal
regions. The relationship between the layers of signals at open chromatin
regions is then learned using a Hidden Markov Model and utilized for predicting
open chromatin. HMMRATAC is capable of outperforming similar methods in
identifying chromatin structure and transcription factor binding sites. 
[@hmmratac2019tarbell] 

A typical analysis pipline begins with aligning sequencing reads to a reference
genome, then using HMMRATAC to identify of accessible regions or "peaks" in the
chromatin. Next motif enrichment can be performed with MEME, footprint
identificaiton with CENTIPEDE, or differential analysis using Diffbind. Quality
control measurments can take place after each of these steps using ATACseqQC.

This Bioconductor package has been made to utilize the original Java program for
HMMRATAC and make it availiable for use within Bioconductor.

# Installation
```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install('HMMRATAC')
```

# Example Pipeline

The rest of this vignette will detail how to use this package and others 
for the purpose of performing analysis on an ATACSeq sample.

We start by first obtaining 

## Generate necessary files using RSamtools

To begin, it is assumed that the user has a BAM file generater from an alligner.
For this example, we will use a file included with the package.
```{r getBAM}
library(Rsamtools)
bam <- system.file("extdata", "65283_treat_rep1.chr22.bam", package="HMMRATAC",
                   mustwork = TRUE)
```
We can use this to generate the remaining two files needed to run HMMRATAC using
Rsamtools. Our next step is to create an index file. Rsamtools can do this with
a sorted BAM file. We generate an index file as follow:
```{r getIndex}
## Generate sorted bam file and save to a temporary location.
sortBam(file = bam, destination = tempfile())

## Create the index file.
index <- indexBam(file = bam)
```
The last file that we will generate is the `genome.info` file. This files shows

```{r getGenomeInfo}
## Get list of headers.
genome <- scanBamHeader(file = bam)

## Extract targets and filter list
targets <- genome[bam]$targets
idx <- grepl("\\.[0-9]+$", names(targets))
genome_info <- targets[!idx]
genome_info <- data.frame(genome_info)

## Write table.
genome <- tempfile()
write.table(genome_info, genome, col.names = FALSE, quote = FALSE)
```

## ATACseqQC

After performing the initial alignment, the quality of the alignment can be
assesed using ATACseqQC

```{r libraryCOmplexity}
library(ATACseqQC)
estimateLibComplexity(readsDupFreq(bam))
```

```{r fragSizeDistribution}
fragSize <- fragSizeDist(bam, "Frag Sizes")
```

Footprint identification can also be performed with ATACseqQC.
```{r footprintID}
library(MotifDb)
CTCF <- query(MotifDb, c("CTCF"))
CTCF <- as.list(CTCF)

sigs <- factorFootprints(shiftedBamfile, pfm=CTCF[[1]], 
                         genome=genome,
                         min.score="90%", seqlev=seqlev,
                         upstream=100, downstream=100)

featureAlignedHeatmap(sigs$signal, 
                      feature.gr=reCenterPeaks(sigs$bindingSites,
                                               width=200+width(sigs$bindingSites[1])), 
                      annoMcols="score",
                      sortBy="score",
                      n.tile=ncol(sigs$signal[[1]]))
```

```{r 

## HMMRATAC

First off, HMMRATAC requires a large allocation of memory to run that rJava
is not able to initially allocate to it. The amount of memory needed varies by
the size of the data being modeled. For this example, we allocate 8GB of memory
before loading the HMMRATAC package:
```{r loadHMMRATAC}
options(java.parameters = "-Xmx8000m")
library(HMMRATAC)
```
Using the original bam file as well as the index and genome info files that we
generated, we can run the HMMRATAC package's main function.

```{r runHMMRATAC}
HMMRATAC(bam = bam, index = index, genome = genome)
```

### HMMR output files

The following files are generated in an output folder from a run of HMMRATAC:

- Name.log

This is a log file generated as HMMRATAC runs. It contains all inputted
arguments, the results of the fragment distribution EM algorithm, and various
status updates as HMMRATAC progresses, as well as a textual description of the
generated model. When finished, it also reports the total amount of actual
(real-world) time it took for HMMRATAC to run.

- Name.bedgraph

This is the genome-wide state annotation file created if the --bedgraph option
is set to true. It is a four column, tab-delimited file, with the following
fields:

  -  Chromosome Name

  -  Region Start (zero-based)

  -  Region Stop (zero-based)

  -  State annotation. This represents what state the particular region is
assigned to, corresponding to the created model. The state name is prefixed
with the letter “E”, to make extractions easier. For instance, to retrieve all
regions that were classified as state 1, type:

  -  grep E1 Name.bedgraph > State1_regions.bed

- Name.model

This file contains the model that was generated by HMMRATAC and was used to
decode the genome. It is always recommended to check this model after running
HMMRATAC. Please note that this file is a binary file that can only be read by
HMMRATAC. To see the actual textual description of the model, see the log file.
If HMMRATAC runs to completion but produces poor results, it is usually the
result of a faulty model. For instance, if the model shows “NaN” for all of the
parameters, such as transition or emission probabilities, this usually means
that something went wrong with either the choice of training regions or
generating the signal tracks. This can occur with too low or too high of a
fold-change range for choosing training regions, if the BED file of training
regions was faulty or if certain signals need to be trimmed off (due to small
numbers of larger fragments). Generally speaking, this poor model occurs when
there is not enough separation in the signal tracks between the states. Common
fixes include, trimming the signal tracks (option --trim), changing the number
of states (option -k), choosing a different fold change range for training site
identification (option -u and -l) or changing the list of previously annotated
training regions (option -t). Additionally, checking the model ensures that the
predicted model is created. Although rare, it is possible for a different state
to be better indicative of the open state, than is normally the case (HMMRATAC
sorts the model so the last or highest numbered state should be the open state).
If this happens, then the resulting peak file may be incorrect. In such cases,
it is recommended to report a bedgraph file and extract the proper state (of
certain lengths and with the accompanying flanking regions) and manually create
a peak BED file. We have tested numerous ATAC-seq datasets, using our default
parameters and we generally don’t encounter problems, unless we radically change
those settings.

- Name_peaks.gappedPeak

This is the standard peak file reported by HMMRATAC by default (when the -p ,
--peaks option is set to true). The first line in the file is a track line for
genome browsers. After that, it is a 15 column, tab-delimited file containg the
following fields:

    - Chromosome Name

    - Peak Start (zero-based). This is the beginning of the regulatory region that
HMMRATAC identifies as a peak. This includes the flanking nucleosome regions.
Therefore, this position represents the start of the upstream nucleosome. If
there is no upstream nucleosome, this position represents the start of the open
state.

    - Peak Stop (zero-based). This is the end of the regulatory region that
HMMRATAC identifies as a peak. This includes the flanking nucleosome regions.
Therefore, this position represents the end of the downstream nucleosome. If
there is no downstream nucleosome, this position represents the end of the open
state.

    - Peak Name. Unique name for the peak, in the format “Peak_#”. For the high
coverage regions that are excluded from Viterbi decoding (described in
Section 2, under the -z , --zscore option), the name is in the format
“HighCoveragePeak_#”. This allows the high coverage peaks to be easily
identified and extracted.

    - Unused, denoted with “.”

    - Unused, denoted with “.”

    - Open state start (zero-based). This is the genomic position where the open
state begins. It is therefore possible to use only the open regions, rather than
the regulatory regions that HMMRATAC reports by default.

    - Open state stop (zero-based). This is the genomic position where the open
state ends. It is therefore possible to use only the open regions, rather than
the regulatory regions that HMMRATAC reports by default.

    - Color code for display. Set to 255,0,0. This would create a dark-blue color
on a genome browser.

    - The number of sub-regions in the peak region. For standard peaks this is set
to 3, indicating the open state region and the two flanking nucleosome state
regions. For high coverage peaks, this is set to 1. Peaks that don’t have
upstream and/or downstream nucleosomes will have different values as well.

    - Comma separated list of sub-region lengths. The first and last values are
always set to one, making visualization easier. The middle value is the length
of the open state region. For high coverage peaks, this is only a single value,
denoting the length of the total region. If the peak lacks upstream and/or
downstream nucleosomes, these values will reflect that.

    - Comma separated list of the (zero-based) starts of each sub-region, relative
to the total region’s start. For instance, the first value is always 0, meaning
0 bp from the region start (column 2). The second value is the distance from the
region start (column 2) to the open region start (column 7), etc.

    - Peak Score. This is the score for the peak, as determined by the scoring
system option described in section 2 (option --score).

    - Unused, denoted as -1.

    - Unused, denoted as -1.

- Name_summits.bed

This is a BED file containing the summits of each HMMRATAC peak. Summits are
determined by finding the position within the open state region, whose
Gaussian-smoothed read coverage is the maximum over the entire region. This is
only outputted if a peak file is also outputted. If motif analysis is being
conducted, it is recommended to use these summits as the points of interest, as
the summits tend to be better indicative of TF binding as opposed to peak
centers. This is a four column, tab-delimited BED file containing the following
fields:

    - Chromosome Name

    - Summit Start (zero-based)

    - Summit Stop (zero-based)

    - Summit Name. This is the same name as the corresponding Peak name used in
the peak file (column 4 of the Name_peaks.gappedPeak file).

    - Peak Score. Same score as in the .gappedPeak file.

## DiffBind

## ChIPseeker

## Gviz

# Session Info
```{r sessionInfo}
sessionInfo()
```
